nvflare/job_config/base_app_config.py

```diff
@@ -50,17 +50,23 @@ def add_task_data_filter(self, tasks: List[str], filter: Filter):
     def add_task_result_filter(self, tasks: List[str], filter: Filter):
         self._add_task_filter(tasks, filter, self.task_result_filters)
 
-    def add_ext_script(self, ext_script: str):
+    def add_ext_script(self, ext_script: str, package_path: str = None):
```

Comment:
Can this be done using PYTHONPATH directly rather than user having to provide a package_path? cc @yanchengnv

Reply:
PYTHONPATH could have multiple matches, which may cause confusion, and cause the wrong codes generated.

Reply:
If ext_script is an absolute dir, then it can be two cases:
- The file contains main(), and the file does not belong to any package
- The file is part of a package.

In the 2nd case, I think we should use PYTHONPATH to determine what the file's relative dir should be.
For example, if the file name is /a/b/c/xyz.py, and the PYTHONPATH contains /a/b, then the file's relative dir should be c/xyz.py.

If, as Yuhong said, the PYTHONPATH also contains /a/b/c, then in theory, you could reference the xyz with "c.xyz" or "xyz".  But not sure this is a common case though.

If we really want to make it identical to user's env, perhaps, we could copy everything verbatim and simply replace "/" with "custom" in the job folder, and then add user's PYTHONPATH to our PYTHONPATH (with / substitued with custom).

In this example, the file /a/b/c/xyz.py will be copied into job folder:  custom/a/b/c/xyz.py.
If the user's PYTHONPATH is /a/b:/a/b/c, then when running on FL server/client, our PYTHONPATH will have: 
<job_dir>/custom/a/b:<job_dir>/custom/a/b/c.

This way, if the user's code has:

import xyz
or 
import c.xyz

It will continue to work.

Reply:
In the case this is a simple script, yes, the package_path may not make too much difference. However, if the script has another import from another source code file, then it's much difference case. The import could event use relative path as well.

When job creation API generating the codes, it needs to parse all the source codes to find all the dependent codes to be included as well.

Reply:
Then this sounds like very deep water, if we have to parse source code. How is it done? What if there are conditional imports? This is what the python interpreter does.

Reply:
Here's the codes to parse the source codes, and find all the dependent codes from the import. Conditional imports will be included as well.

https://github.com/NVIDIA/NVFlare/blob/2a71c61ed0774433431e8c920799aca39ac5a874/nvflare/job_config/fed_job_config.py#L189-L206

Reply:
This doesn't handle all cases. For example:

from \
    nvflare.app_common.tie.defs \
    import \
    Constant as TieConstant

It's almost like that we are writing a small lex parser here.

Reply:
yeah, the import are required to be in the same line.

Reply:
Added handle for the line break of continuous lines support, added unit tests.

Also removed the requirement of external absolute path script to provide the package path. Use the longest matching path in the PYTHONPATH instead.